dart/collision/CollisionGroup.h

```diff
@@ -0,0 +1,210 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COLLISION_COLLISIONGROUP_H_
+#define DART_COLLISION_COLLISIONGROUP_H_
+
+#include <vector>
+#include "dart/collision/SmartPointer.h"
+#include "dart/collision/Option.h"
+#include "dart/collision/Result.h"
+#include "dart/dynamics/SmartPointer.h"
+
+namespace dart {
+namespace collision {
+
+class CollisionGroup
+{
+public:
+
+  /// Constructor
+  CollisionGroup(const CollisionDetectorPtr& collisionDetector);
+
+  /// Destructor
+  virtual ~CollisionGroup();
+
+  /// Return collision detection engine associated with this CollisionGroup
+  CollisionDetector* getCollisionDetector() const;
+
+  /// Register a ShapeFrame to this CollisionGroup
+  void registerShapeFrame(const dynamics::ShapeFrame* shapeFrame);
+
+  /// Register ShapeFrames to this CollisionGroup
+  void registerShapeFrames(
+      const std::vector<const dynamics::ShapeFrame*>& shapeFrames);
+
+  /// Register all the ShapeFrames in the other CollisionGroups to this
+  /// CollisionGroup
+  template <typename First, typename... Others>
+  void registerShapeFrames(const First* first, const Others*... others);
+
+  /// Do nothing. This function is for terminating the variadic template
+  /// function template<typename...> registerShapeFrames().
+  void registerShapeFrames();
+
+  /// As as registerShapeFrame()
+  void registerShapeFramesOf(const dynamics::ShapeFrame* shapeFrame);
+
+  /// As as registerShapeFrames()
+  void registerShapeFramesOf(
+      const std::vector<const dynamics::ShapeFrame*>& shapeFrames);
+
+  /// Register all the ShapeFrames in the other CollisionGroup to this
+  /// CollisionGroup
+  void registerShapeFramesOf(const CollisionGroup* other);
+
+  /// Register all the BodyNode's ShapeFrames having CollisionAddon to this
+  /// CollisionGroup
+  void registerShapeFramesOf(const dynamics::BodyNode* bodyNode);
```

Comment:
I am confused about the semantic of these helper functions: If I call this function on a `BodyNode *`, then add a `ShapeFrame` to it, will that change be propagated to this group?

I don't believe that is the case. If so, we should clearly document this in the comment.